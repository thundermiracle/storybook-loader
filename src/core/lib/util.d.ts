import React from 'react';

declare function basename(filePath: string, regExpForRemove?: RegExp): string;
declare function foldername(filePath: string): string;
declare function getRegExpFromRequireContext(requestContext: any): RegExp;
declare function isFileNameCorrect(
  filePath: string,
  includeRegExp: RegExp,
  excludeRegExp?: RegExp,
): boolean;
declare function isFilePathCorrect(
  filePath: string,
  includeRegExp: RegExp,
  excludeRegExp?: RegExp,
): boolean;

type FunctionListElement =
  | string
  | {
      funcName: string;
      params?: Array<string>;
    };
declare function applySubFuncs(baseObj: any, funcList?: Array<FunctionListElement>): void;
declare function isRequireContextRegExpPassed(requestContext: any): boolean;
declare function flattenContentObj(
  contentObj: object,
  containGroupFolder?: boolean,
): Array<Array<string>>;
declare function getComponent(fileContent: NodeModule): React.Component<any, any>;
declare function applyFuncList(
  component: React.Component<any, any>,
  funcList?: Function[],
): React.Component<any, any>;
declare function unaryFunc(func: Function, params?: Array<any>): Function;
declare function toList(elem: any): Array<any>;
declare function toReact(Component: React.Component<any, any>): JSX.Element;

export {
  FunctionListElement,
  basename,
  foldername,
  getRegExpFromRequireContext,
  isFileNameCorrect,
  isFilePathCorrect,
  applySubFuncs,
  isRequireContextRegExpPassed,
  flattenContentObj,
  getComponent,
  applyFuncList,
  unaryFunc,
  toList,
  toReact,
};
