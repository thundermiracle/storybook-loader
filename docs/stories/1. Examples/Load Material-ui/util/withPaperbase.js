import React from 'react';
import Paperbase from './paperbase';

export default (Component) => {
  const WithPaperbase = (props) => (
    <Paperbase>
      <Component {...props} />
    </Paperbase>
  );

  return WithPaperbase;
}
