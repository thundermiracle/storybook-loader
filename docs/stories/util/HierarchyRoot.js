export default (function Hierarchy() {
  return {
    Markdown: 'Examples|Load Markdown/',
    MaterialUI: 'Examples|Load Material-UI/',
    ReactComponent: 'Examples|Load React Component/',
    ReactComponentWithNotes: 'Examples|Load React Component WithNotes/',
    RootMarkdown: 'Examples|Load Markdown in root/',
  };
})();
