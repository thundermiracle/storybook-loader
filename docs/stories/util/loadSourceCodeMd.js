import { loadMDStories } from 'storybook-loader';
import { doc } from 'storybook-readme';

const options = {
  contentFuncList: [
    doc,
  ],
  ignoreDotFolder: false,
  dotFolderName: '',
  storySubFuncList: [
    [
      'addParameters',
      [{ options: { showPanel: false } }],
    ],
  ],
};

export default function loadSourceCodeMd(req, userOptions = {}) {
  loadMDStories(req, { ...options, ...userOptions });
}
