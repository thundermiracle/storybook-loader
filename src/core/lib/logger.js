const logger = console;

const LogLevel = {
  silent: 0,
  error: 1,
  warn: 2,
  info: 3,
  debug: 4,
};

function loggerFunctionMaker(logLevel) {
  return function(type) {
    return function(...params) {
      if (logLevel >= LogLevel[type]) {
        logger[type](...params);
      }
    };
  };
}

function CustomLogger(logLevel) {
  const loggerTypes = ['error', 'warn', 'info', 'debug'];

  return loggerTypes.reduce((funcObj, type) => {
    funcObj[type] = loggerFunctionMaker(logLevel)(type);
    return funcObj;
  }, {});
}

let level = LogLevel[process.env.LOG_LEVEL];
if (level == null) {
  level = process.env.NODE_ENV === 'production' ? LogLevel.error : LogLevel.debug;
}

const log = new CustomLogger(level);

export { log as default, LogLevel, CustomLogger };
