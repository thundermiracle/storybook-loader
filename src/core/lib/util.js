/* eslint-disable no-unused-expressions */
import path from 'path';
import { toPairs, __, pipe, curry } from 'ramda';

import log from './logger';

// const allExtRegExp = /\.[0-9a-z]+$/i;
const allExtRegExp = /(.*?)/;

/**
 * Return the basename of the file
 *
 * @param p the path to evaluate
 * @param regExpForRemove optionally, regex for remove
 */
function basename(filePath, regExpForRemove = null) {
  let fileName = path.basename(filePath);
  if (regExpForRemove) {
    fileName = fileName.replace(regExpForRemove, '');
  }
  return fileName;
}

/**
 * Return the folder name.
 * If folder is relative, dot will be retrieved.
 * @example ./path.1/path2/filename  =>  path.1/path2
 * @example /path.1/path2/filename  =>  path.1/path2
 *
 * @param p the path to evaluate
 */
function foldername(filePath) {
  return path.dirname(filePath).replace(/^.?\//, '');
}

function getRegExpStrFromRequireContext(req) {
  const id = req.id;
  if (id == null || typeof id !== 'string') {
    return null;
  }

  const regExpStr = id.split(' ').pop();
  if (regExpStr === '') {
    return null;
  }

  return regExpStr;
}

/**
 * Return RegExp string in require.context instance
 *
 * @param {*require.context} req webpack's require.context instance
 */
function getRegExpFromRequireContext(req) {
  const regExpStr = getRegExpStrFromRequireContext(req);

  return regExpStr != null ? new RegExp(regExpStr) : null;
}

/**
 * Return whether regexp is passed to require.context or not.
 *
 * @param {*require.context} req webpack's require.context instance
 */
function isRequireContextRegExpPassed(req) {
  const regExpFromReq = getRegExpStrFromRequireContext(req);
  return regExpFromReq != null && regExpFromReq !== '^\\.\\/.*$';
}

/**
 * Return true if file's name match RegExp
 *
 * @param {*string} filePath
 * @param {*RegExp} regExp
 */
function isFileNameCorrect(filePath, includeRegExp, excludeRegExp = null) {
  const fileName = path.basename(filePath);
  return (
    (includeRegExp || allExtRegExp).test(fileName) &&
    (excludeRegExp == null ? true : !excludeRegExp.test(fileName))
  );
}

/**
 * Return true if file's full path(include filename) match RegExp
 *
 * @param {*string} filePath
 * @param {*RegExp} regExp
 */
function isFilePathCorrect(filePath, includeRegExp, excludeRegExp = null) {
  return (
    (includeRegExp || allExtRegExp).test(filePath) &&
    (excludeRegExp == null ? true : !excludeRegExp.test(filePath))
  );
}

/**
 * Apply all functions in funcObj to baseObj.
 * @example funcList: [ "func1", ["func2"], ["func3", [ param1, param2 ]];
 * will call -- baseObj.func1() -> baseObj.func2() -> baseObj.func3(param1, param2);
 *
 * @param {*instance} baseObj main function
 * @param {*object} funcObj sub function object
 */
function applySubFuncs(baseObj, funcList = []) {
  funcList.forEach(funcInfo => {
    if (typeof funcInfo === 'string') {
      // only sub function
      baseObj[funcInfo] && baseObj[funcInfo]();
    } else {
      const [funcName, params] = funcInfo;
      if (params == null) {
        // run sub function without params
        baseObj[funcName] && baseObj[funcName]();
        return;
      }

      if (!Array.isArray(params)) {
        log.error(
          `function [${funcName}]'s parameters in storySubFuncList MUST be an array.`,
        );
        return;
      }

      baseObj[funcName] && baseObj[funcName](...params);
    }
  });
}

/**
 * Return flattened array of contentObj
 * @example { a1: '111', b1: '222' }  ->  [['', [['a1', '111'], ['b1', '222']]]]
 * @example { fa: { a1: '111', a2: '112'}, fb: { b1: '222'}}  ->  [["fa", [["a1", "111"], ["a2", "112"]]], ["fb", [["b1", "222"]]]]
 *
 * @param contentObj object of filecontents loaded by loader.js
 * @param containGroupFolder boolean
 */
function flattenContentObj(contentObj, containGroupFolder = false) {
  // flatten folder layer
  let folderBaseContentList;
  if (containGroupFolder) {
    folderBaseContentList = toPairs(contentObj).map(([folderName, fileContentObj]) => [
      folderName,
      toPairs(fileContentObj),
    ]);
  } else {
    // if folder layer is not exist, add a space folder
    folderBaseContentList = [['', toPairs(contentObj)]];
  }

  return folderBaseContentList;
}

/**
 * Return react Component(function) from fileContent
 *
 * @param fileContent instance of req(fileName)
 */
function getComponent(fileContent) {
  if (typeof fileContent === 'object' && typeof fileContent.default === 'function') {
    return fileContent.default;
  }
  return fileContent;
}

// /**
//  * Return composed purefunction and skip setting the first parameter
//  *
//  * @param funcList
//  */
// function composePureFuncSkipFirstParam(funcList) {
//   const allFuncs = funcList.map(({ func, params }) => {
//     if (params == null) {
//       return func;
//     }

//     return curry(func)(__, ...params);
//   });

//   return pipe(...allFuncs);
// }

/**
 * Return component applied all functions in funcList, funcList MUST be pure
 * @example funcList: [hoc1, hoc2]
 *  will return hoc2(hoc1(component))
 *
 * @param component target Component
 * @param funcList function list
 */
function applyFuncList(component, funcList = []) {
  if (funcList.length === 0) {
    return component;
  }

  const composedFuncs = pipe(...funcList);

  return composedFuncs(component);
}

/**
 * Return new function with setting all params from second place.
 * @example (function myName(name, msg1, msg2) {}, params=['hello', 'world'])  =>
 *   myName(name, 'hello', 'world');
 *
 * @param func
 * @param params parameter array
 */
function unaryFunc(func, params = []) {
  if (params == null || params.length === 0) {
    return func;
  }

  return curry(func)(__, ...params);
}

/**
 * Change input element to list if it's not
 *
 * @param {any} elem
 */
function toList(elem) {
  if (Array.isArray(elem)) return elem;

  return [elem];
}

function toReact(Component) {
  const toReactFunc = props => <Component {...props} />;
  return toReactFunc;
}

export {
  basename,
  foldername,
  getRegExpFromRequireContext,
  isFileNameCorrect,
  isFilePathCorrect,
  applySubFuncs,
  isRequireContextRegExpPassed,
  flattenContentObj,
  getComponent,
  applyFuncList,
  unaryFunc,
  toList,
  toReact,
};
