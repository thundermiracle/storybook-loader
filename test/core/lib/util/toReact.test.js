import renderer from 'react-test-renderer';

import { toReact } from 'core/lib/util';

test('[string]input is string', () => {
  const originStr = 'origin';
  const result = toReact(originStr);
  expect(typeof result).toEqual('function');
});

test('React component can be rendered', () => {
  const divReact = require('test/mocks/DivReact').default;
  const ToReactComponent = toReact(divReact);

  const component = renderer.create(
    <ToReactComponent content="First" />,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
