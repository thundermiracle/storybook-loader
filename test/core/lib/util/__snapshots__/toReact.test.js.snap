// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React component can be rendered 1`] = `
<div>
  First
</div>
`;
